assets/js/base/components/checkbox-list/index.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * External dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+import PropTypes from 'prop-types';
+import {
+	Fragment,
+	useCallback,
+	useMemo,
+	useState,
+	useEffect,
+} from '@wordpress/element';
+import classNames from 'classnames';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+
+/**
+ * Component used to show a list of checkboxes in a group.
+ */
+const CheckboxList = ( {
+	className,
+	onChange = () => {},
+	options = [],
+	isLoading = false,
+	limit = 10,
+} ) => {
+	// Holds all checked options.
+	const [ checked, setChecked ] = useState( [] );
+	const [ showExpanded, setShowExpanded ] = useState( false );
+
+	useEffect( () => {
+		onChange( checked );
+	}, [ checked ] );
+
+	const placeholder = useMemo( () => {
+		return (
+			<Fragment>
+				{ [ ...Array( 5 ) ].map( ( x, i ) => (
+					<li
+						key={ i }
+						style={ {
+							/* stylelint-disable */
+							width: Math.floor( Math.random() * 75 ) + 25 + '%',
+						} }
+					/>
+				) ) }
+			</Fragment>
+		);
+	}, [] );
+
+	const onCheckboxChange = useCallback(
+		( event ) => {
+			const isChecked = event.target.checked;
+			const checkedValue = event.target.value;
+			const newChecked = checked.filter(
+				( value ) => value !== checkedValue
+			);
+
+			if ( isChecked ) {
+				newChecked.push( checkedValue );
+				newChecked.sort();
+			}
+
+			setChecked( newChecked );
+		},
+		[ checked ]
+	);
+
+	const renderedShowMore = useMemo( () => {
+		const optionCount = options.length;
+		return (
+			! showExpanded && (
+				<li className="show-more">
+					<button
+						onClick={ () => {
+							setShowExpanded( true );
+						} }
+						aria-expanded={ false }
+						aria-label={ sprintf(
+							__(
+								'Show %s more options',
+								'woo-gutenberg-products-block'
+							),
+							optionCount - limit
+						) }
+					>
+						{ // translators: %s number of options to reveal.
+						sprintf(
+							__(
+								'Show %s more',
+								'woo-gutenberg-products-block'
+							),
+							optionCount - limit
+						) }
+					</button>
+				</li>
+			)
+		);
+	}, [ options, limit, showExpanded ] );
+
+	const renderedShowLess = useMemo( () => {
+		return (
+			showExpanded && (
+				<li className="show-less">
+					<button
+						onClick={ () => {
+							setShowExpanded( false );
+						} }
+						aria-expanded={ true }
+						aria-label={ __(
+							'Show less options',
+							'woo-gutenberg-products-block'
+						) }
+					>
+						{ __( 'Show less', 'woo-gutenberg-products-block' ) }
+					</button>
+				</li>
+			)
+		);
+	}, [ showExpanded ] );
+
+	const renderedOptions = useMemo( () => {
+		// Truncate options if > the limit + 5.
+		const optionCount = options.length;
+		const shouldTruncateOptions = optionCount > limit + 5;
+		return (
+			<Fragment>
+				{ options.map( ( option, index ) => (
+					<Fragment key={ option.key }>
```

Comment:
This fragment wraps 2 sibling `<li>` elements so is needed here?

Reply:
You're right there is _one_ Fragment that could remain.  However there's a couple things:

1. The inner level fragment should still not be needed.
2. The key should be on the li items, not the Fragment.  The problem with the Fragment having the key is that the "show more options" toggle will get re-rendered everytime the inner list re-renders even if it's not necessary.

The above points might not see important in the context of this particular usage, but it's pattern that could be more problematic in other contexts if it becomes a habit :). So I'm just promoting best practices here.

This code could get changed to:

```jsx
useMemo( () => {
	// Truncate options if > the limit + 5.
	const optionCount = options.length;
	const shouldTruncateOptions = optionCount > limit + 5;
	return (
		<Fragment>
			{ options.map( ( option, index ) => (
				<li
					key={ option.key }
					{ ...shouldTruncateOptions &&
						! showExpanded &&
						index >= limit && { hidden: true } }
				>
					<input
						type="checkbox"
						id={ option.key }
						value={ option.key }
						onChange={ onCheckboxChange }
						checked={ checked.includes( option.key ) }
					/>
					<label htmlFor={ option.key }>{ option.label }</label>
				</li>
			) ) }
			{ shouldTruncateOptions &&
				options.length > limit &&
				renderedShowMore }
			{ shouldTruncateOptions && renderedShowLess }
		</Fragment>
	);
}, [
	options,
	checked,
	showExpanded,
	limit,
	onCheckboxChange,
	renderedShowLess,
	renderedShowMore,
] );
```

Even better though, would be to handle the `renderedShowMore` and `renderedShowLess` outside of this memoized function.  I'd combine the `renderedShowMore` and `renderedShowLess` into one memoized function that spits out either of those or null based on the conditions. Then you could get rid of the last unnecessary fragment and end up with something like this:

```jsx
<ul className={ classes }>
    { isLoading ? placeholder : renderedOptions  }
    { renderedToggleOptions }
</ul>
```

Reply:
Hmm but by having the show more after the list items, if you tab through (with keyboard) when it expands you're at the bottom of the list  rather than from the point of expansion.. that was my justification for having it appear inline. Originally I did it similar to you but only rendering visible list items. You were against that particular solution also.

Reply:
🤔 hmm... I see your point.  I tried both behaviours and what you have definitely works better from an a11y pov. Let's roll with it.

Reply:
>  The problem with the Fragment having the key is that the "show more options" toggle will get re-rendered everytime the inner list re-renders even if it's not necessary.

Did we work around this by adding a key to the show more/show less li?

Reply:
> Did we work around this by adding a key to the show more/show less li?

Hmm... I don't think so (not sure honestly), but regardless I think in this case any accessibility improvements make it worth potential extra rendering for the time being (until we can think of a better pattern).